test_quality_of_service/test/test_liveliness.cpp

```diff
@@ -0,0 +1,129 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include <chrono>
+#include <exception>
+#include <iostream>
+#include <memory>
+#include <stack>
+#include <string>
+#include <tuple>
+
+#include "gtest/gtest.h"
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp/executors/single_threaded_executor.hpp"
+
+#include "std_msgs/msg/string.hpp"
+
+#include "test_quality_of_service/qos_test_publisher.hpp"
+#include "test_quality_of_service/qos_test_subscriber.hpp"
+#include "test_quality_of_service/qos_utilities.hpp"
+
+using namespace std::chrono_literals;
+
+/// Test Automatic Liveliness with a single publishing node and single subscriber node
+TEST_F(QosRclcppTestFixture, test_automatic_liveliness_changed) {
+  const std::chrono::milliseconds max_test_length = 8s;
+  const std::chrono::milliseconds kill_publisher_after = 2s;
+  const std::chrono::milliseconds publish_period = 200ms;
+  int number_of_published_messages = 0;
+  const std::chrono::milliseconds lease_duration = 1s;
+
+  int total_number_of_liveliness_events = 0;
+
+  // define qos profile
+  rclcpp::QoS qos_profile(10);
+  qos_profile.liveliness(RMW_QOS_POLICY_LIVELINESS_AUTOMATIC);
+  qos_profile.liveliness_lease_duration(lease_duration);
+
+  // subscription options
+  rclcpp::SubscriptionOptions subscriber_options;
+  // subscriber Liveliness callback event
+  subscriber_options.event_callbacks.liveliness_callback =
+    [&total_number_of_liveliness_events](rclcpp::QOSLivelinessChangedInfo & event) -> void
+    {
+      RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "QOSLivelinessChangedInfo callback");
```

Comment:
Same here, might just want to avoid `rclcpp::get_logger()` in all cases possible, it's kind of a "bad code smell".

Reply:
+1